File: src/rizzo_flow/decisions.py
"""Pure, deterministic decoding of categorical logits into typed results."""

import math
from dataclasses import dataclass

from .responses import (
    BooleanAnswer,
    ChoiceAnswer,
    NumericAnswer,
    ScoreAnswer,
    Statistics,
    TypedAnswer,
    Uncertainty,
)
from .schema import BooleanQuestion, ChoiceQuestion, NumericQuestion, Question, ScoreQuestion

UNKNOWN = "__insufficient__"
BELOW = "__below_range__"
ABOVE = "__above_range__"


@dataclass(frozen=True)
class Candidate:
    id: str
    description: str
    value: float | None = None


def candidates(question: Question) -> list[Candidate]:
    if isinstance(question, BooleanQuestion):
        result = [
            Candidate("false", question.false_description, 0),
            Candidate("true", question.true_description, 1),
        ]
    elif isinstance(question, ChoiceQuestion):
        result = [Candidate(o.id, o.description) for o in question.options]
    elif isinstance(question, ScoreQuestion):
        result = [Candidate(str(i), level, i) for i, level in enumerate(question.levels)]
    elif isinstance(question, NumericQuestion):
        result = [
            Candidate(
                str(i), f"Approximately {a.value:g} {question.unit}: {a.description}", a.value
            )
            for i, a in enumerate(question.anchors)
        ]
        result += [
            Candidate(BELOW, f"The value is below {question.anchors[0].value:g} {question.unit}."),
            Candidate(ABOVE, f"The value is above {question.anchors[-1].value:g} {question.unit}."),
        ]
    else:
        raise TypeError("Unsupported question")
    if question.policy.allow_abstain:
        result.append(
            Candidate(
                UNKNOWN,
                "Cannot determine the answer: the required information is not provided "
                "or is contradictory. A known value outside the numeric range is not "
                "missing information.",
            )
        )
    return result


def softmax(logits: list[float], temperature: float = 1.0) -> list[float]:
    if not math.isfinite(temperature) or temperature <= 0:
        raise ValueError("Temperature must be finite and positive")
    if len(logits) < 2 or not all(math.isfinite(x) for x in logits):
        raise ValueError("At least two finite logits are required")
    maximum = max(logits)
    weights = [math.exp((x - maximum) / temperature) for x in logits]
    total = sum(weights)
    return [x / total for x in weights]


def summarize(values: list[float], probabilities: list[float]) -> Statistics:
    mean = math.fsum(v * p for v, p in zip(values, probabilities, strict=True))
    variance = math.fsum(p * (v - mean) ** 2 for v, p in zip(values, probabilities, strict=True))

    def quantile(q: float) -> float:
        cumulative = 0.0
        for v, p in zip(values, probabilities, strict=True):
            cumulative += p
            if cumulative >= q:
                return v
        return values[-1]

    return Statistics(
        mean=mean,
        stddev=math.sqrt(variance),
        median=quantile(0.5),
        anchor_quantiles={"p10": quantile(0.1), "p90": quantile(0.9)},
    )


def decode(
    question: Question,
    logits: list[float],
    *,
    prompt_sha256: str,
    input_tokens: int,
    temperature: float = 1.0,
) -> TypedAnswer:
    """Turn the answer-letter logits into the typed answer this question asked for.

    `prompt_sha256` and `input_tokens` describe the prompt these logits came from. They
    are arguments rather than something the caller patches in afterwards, so an answer is
    complete the moment it exists.
    """
    choices = candidates(question)
    if len(logits) != len(choices):
        raise ValueError("Logit count does not match the declared candidates")
    ps = softmax(logits, temperature)
    distribution = {c.id: p for c, p in zip(choices, ps, strict=True)}
    unavailable_ids = {UNKNOWN, BELOW, ABOVE}
    valid = [(c, p) for c, p in zip(choices, ps, strict=True) if c.id not in unavailable_ids]
    unavailable = math.fsum(p for c, p in zip(choices, ps, strict=True) if c.id in unavailable_ids)
    available = math.fsum(p for _, p in valid)
    winner = choices[max(range(len(ps)), key=ps.__getitem__)].id
    top = max(ps)
    entropy = -math.fsum(p * math.log(p) for p in ps if p > 0)
    status = "ok"
    if winner in unavailable_ids or unavailable >= question.policy.max_unavailable_probability:
        status = (
            "out_of_range"
            if distribution.get(BELOW, 0) + distribution.get(ABOVE, 0)
            > distribution.get(UNKNOWN, 0)
            else "insufficient_evidence"
        )
    elif top < question.policy.min_top_probability:
        status = "uncertain"
    shared = {
        "status": status,
        "probabilities": distribution,
        "option_logits": {c.id: x for c, x in zip(choices, logits, strict=True)},
        "legend": {c.id: c.description for c in choices},
        "uncertainty": Uncertainty(
            top_probability=top,
            entropy_nats=entropy,
            concentration=max(0.0, min(1.0, 1 - entropy / math.log(len(ps)))),
            unavailable_probability=unavailable,
        ),
        "probability_status": "uncalibrated_conditional_option_scores"
        if temperature == 1
        else "temperature_scaled_requires_held_out_validation",
        "temperature": temperature,
        "prompt_sha256": prompt_sha256,
        "input_tokens": input_tokens,
    }
    conditional = [p / available for _, p in valid] if available > 0 else None
    if isinstance(question, ChoiceQuestion):
        return ChoiceAnswer(**shared, type="choice", choice=winner if status == "ok" else None)
    if isinstance(question, BooleanQuestion):
        return BooleanAnswer(
            **shared,
            type="boolean",
            value=(winner == "true") if status == "ok" else None,
            probability_true_given_available=conditional[1] if conditional else None,
        )
    values = [c.value for c, _ in valid]
    stats = summarize(values, conditional) if conditional and status == "ok" else None
    scale = {
        "statistics_given_available": stats,
        "values": {c.id: c.value for c, _ in valid},
        "support": [values[0], values[-1]],
    }
    if isinstance(question, NumericQuestion):
        return NumericAnswer(
            **shared,
            **scale,
            type="numeric",
            value=stats.mean if stats else None,
            unit=question.unit,
            range_probabilities={"below": distribution[BELOW], "above": distribution[ABOVE]},
        )
    return ScoreAnswer(
        **shared,
        **scale,
        type="score",
        score=stats.mean if stats else None,
        normalized_score=stats.mean / values[-1] if stats else None,
    )

File: rust/src/decisions.rs
//! Pure, deterministic decoding of categorical logits into typed results.
//! Rust port of `decisions.py` (Rizzo Flow, Apache-2.0) — no I/O, no side effects.

use crate::schema::{Question, ValidationError};
use serde_json::{json, Map, Value};

pub const UNKNOWN: &str = "__insufficient__";
pub const BELOW: &str = "__below_range__";
pub const ABOVE: &str = "__above_range__";

pub type Result<T> = std::result::Result<T, ValidationError>;

fn err<T>(msg: impl Into<String>) -> Result<T> {
    Err(ValidationError(msg.into()))
}

#[derive(Debug, Clone, PartialEq)]
pub struct Candidate {
    pub id: String,
    pub description: String,
    pub value: Option<f64>,
}

/// Build the candidate list for a question (options + special abstention/range slots).
pub fn candidates(question: &Question) -> Vec<Candidate> {
    let mut result: Vec<Candidate> = match question {
        Question::Boolean(q) => vec![
            Candidate {
                id: "false".into(),
                description: q.false_description.clone(),
                value: Some(0.0),
            },
            Candidate {
                id: "true".into(),
                description: q.true_description.clone(),
                value: Some(1.0),
            },
        ],
        Question::Choice(q) => q
            .options
            .iter()
            .map(|o| Candidate {
                id: o.id.clone(),
                description: o.description.clone(),
                value: None,
            })
            .collect(),
        Question::Score(q) => q
            .levels
            .iter()
            .enumerate()
            .map(|(i, level)| Candidate {
                id: i.to_string(),
                description: level.clone(),
                value: Some(i as f64),
            })
            .collect(),
        Question::Numeric(q) => {
            let mut list: Vec<Candidate> = q
                .anchors
                .iter()
                .enumerate()
                .map(|(i, a)| Candidate {
                    id: i.to_string(),
                    description: format!("Approximately {} {}: {}", a.value, q.unit, a.description),
                    value: Some(a.value),
                })
                .collect();
            let first = q.anchors.first().map(|a| a.value).unwrap_or(0.0);
            let last = q.anchors.last().map(|a| a.value).unwrap_or(0.0);
            list.push(Candidate {
                id: BELOW.into(),
                description: format!("The value is below {first} {}.", q.unit),
                value: None,
            });
            list.push(Candidate {
                id: ABOVE.into(),
                description: format!("The value is above {last} {}.", q.unit),
                value: None,
            });
            list
        }
    };
    if question.policy().allow_abstain {
        result.push(Candidate {
            id: UNKNOWN.into(),
            description: "Cannot determine the answer: the required information is not provided \
                          or is contradictory. A known value outside the numeric range is not \
                          missing information."
                .into(),
            value: None,
        });
    }
    result
}

/// Numerically stable softmax with optional temperature.
pub fn softmax(logits: &[f64], temperature: f64) -> Result<Vec<f64>> {
    if !temperature.is_finite() || temperature <= 0.0 {
        return err("Temperature must be finite and positive");
    }
    if logits.len() < 2 || !logits.iter().all(|x| x.is_finite()) {
        return err("At least two finite logits are required");
    }
    let maximum = logits.iter().copied().fold(f64::NEG_INFINITY, f64::max);
    let weights: Vec<f64> = logits
        .iter()
        .map(|x| ((x - maximum) / temperature).exp())
        .collect();
    let total: f64 = weights.iter().sum();
    Ok(weights.iter().map(|w| w / total).collect())
}

/// Probability-weighted summary for score/numeric answers.
pub fn summarize(values: &[f64], probabilities: &[f64]) -> Value {
    let mean: f64 = values
        .iter()
        .zip(probabilities)
        .map(|(v, p)| v * p)
        .sum();
    let variance: f64 = probabilities
        .iter()
        .zip(values)
        .map(|(p, v)| p * (v - mean).powi(2))
        .sum();
    let quantile = |q: f64| -> f64 {
        let mut cumulative = 0.0;
        for (v, p) in values.iter().zip(probabilities) {
            cumulative += p;
            if cumulative >= q {
                return *v;
            }
        }
        values.last().copied().unwrap_or(0.0)
    };
    json!({
        "mean": mean,
        "stddev": variance.sqrt(),
        "median": quantile(0.5),
        "anchor_quantiles": {"p10": quantile(0.1), "p90": quantile(0.9)},
    })
}

/// Decode logits (one per candidate) into a typed result — parity with `decisions.decode`.
pub fn decode(question: &Question, logits: &[f64], temperature: f64) -> Result<Value> {
    let choices = candidates(question);
    if logits.len() != choices.len() {
        return err("Logit count does not match the declared candidates");
    }
    let ps = softmax(logits, temperature)?;

    let distribution: Map<String, Value> = choices
        .iter()
        .zip(&ps)
        .map(|(c, p)| (c.id.clone(), json!(p)))
        .collect();
    let unavailable_ids = [UNKNOWN, BELOW, ABOVE];
    let mut valid: Vec<(&Candidate, f64)> = Vec::new();
    let mut unavailable = 0.0_f64;
    for (c, p) in choices.iter().zip(&ps) {
        if unavailable_ids.contains(&c.id.as_str()) {
            unavailable += p;
        } else {
            valid.push((c, *p));
        }
    }
    let available: f64 = valid.iter().map(|(_, p)| p).sum();

    let (winner_idx, top) = ps
        .iter()
        .enumerate()
        .fold((0_usize, f64::NEG_INFINITY), |(bi, bt), (i, p)| {
            if *p > bt {
                (i, *p)
            } else {
                (bi, bt)
            }
        });
    let winner = choices[winner_idx].id.as_str();
    let entropy: f64 = ps.iter().filter(|p| **p > 0.0).map(|p| -p * p.ln()).sum();

    let policy = question.policy();
    let mut status = "ok";
    if unavailable_ids.contains(&winner) || unavailable >= policy.max_unavailable_probability {
        let get = |id: &str| distribution.get(id).and_then(Value::as_f64).unwrap_or(0.0);
        status = if get(BELOW) + get(ABOVE) > get(UNKNOWN) {
            "out_of_range"
        } else {
            "insufficient_evidence"
        };
    } else if top < policy.min_top_probability {
        status = "uncertain";
    }
    let concentration = (1.0 - entropy / (ps.len() as f64).ln()).clamp(0.0, 1.0);
    let probability_status = if temperature == 1.0 {
        "uncalibrated_conditional_option_scores"
    } else {
        "temperature_scaled_requires_held_out_validation"
    };

    let mut result = json!({
        "type": question.type_name(),
        "status": status,
        "probabilities": distribution,
        "option_logits": choices.iter().zip(logits)
            .map(|(c, x)| (c.id.clone(), json!(x)))
            .collect::<Map<String, Value>>(),
        "legend": choices.iter()
            .map(|c| (c.id.clone(), json!(c.description)))
            .collect::<Map<String, Value>>(),
        "uncertainty": {
            "top_probability": top,
            "entropy_nats": entropy,
            "concentration": concentration,
            "unavailable_probability": unavailable,
        },
        "probability_status": probability_status,
        "temperature": temperature,
    });

    let conditional: Option<Vec<f64>> = if available > 0.0 {
        Some(valid.iter().map(|(_, p)| p / available).collect())
    } else {
        None
    };

    match question {
        Question::Choice(_) => {
            result["choice"] = if status == "ok" { json!(winner) } else { Value::Null };
        }
        Question::Boolean(_) => {
            result["value"] = if status == "ok" {
                json!(winner == "true")
            } else {
                Value::Null
            };
            result["probability_true_given_available"] = conditional
                .as_ref()
                .and_then(|c| c.get(1))
                .map(|p| json!(p))
                .unwrap_or(Value::Null);
        }
        _ => {
            let values: Vec<f64> = valid.iter().filter_map(|(c, _)| c.value).collect();
            let stats = match (&conditional, status) {
                (Some(cond), "ok") => Some(summarize(&values, cond)),
                _ => None,
            };
            let mean = stats
                .as_ref()
                .and_then(|s| s.get("mean"))
                .and_then(Value::as_f64);
            result["statistics_given_available"] = stats.clone().unwrap_or(Value::Null);
            result["values"] = Value::Object(
                valid
                    .iter()
                    .map(|(c, _)| (c.id.clone(), json!(c.value)))
                    .collect::<Map<String, Value>>(),
            );
            result["support"] = json!([values.first().copied(), values.last().copied()]);
            match question {
                Question::Numeric(q) => {
                    result["value"] = mean.map(|m| json!(m)).unwrap_or(Value::Null);
                    result["unit"] = json!(q.unit);
                    result["range_probabilities"] = json!({
                        "below": distribution.get(BELOW).cloned().unwrap_or(json!(0.0)),
                        "above": distribution.get(ABOVE).cloned().unwrap_or(json!(0.0)),
                    });
                }
                Question::Score(_) => {
                    result["score"] = mean.map(|m| json!(m)).unwrap_or(Value::Null);
                    result["normalized_score"] = mean
                        .and_then(|m| values.last().map(|last| json!(m / last)))
                        .unwrap_or(Value::Null);
                }
                _ => unreachable!(),
            }
        }
    }
    Ok(result)
}
